// llama-node/dist/index.js
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.ts
var LLM = class {
  constructor(llm) {
    this.llm = new llm();
  }
  load(config) {
    return this.llm.load(config);
  }
  createCompletion(params, callback) {
    return __async(this, null, function* () {
      return this.llm.createCompletion(params, callback);
    });
  }
  getEmbedding(params) {
    return __async(this, null, function* () {
      if (!this.llm.getEmbedding) {
        console.warn("getEmbedding not implemented for current LLM");
        return [];
      } else {
        return this.llm.getEmbedding(params);
      }
    });
  }
  getDefaultEmbeddings(text) {
    return __async(this, null, function* () {
      if (!this.llm.getDefaultEmbedding) {
        console.warn("getDefaultEmbedding not implemented for current LLM");
        return [];
      } else {
        return this.llm.getDefaultEmbedding(text);
      }
    });
  }
  tokenize(content) {
    return __async(this, null, function* () {
      if (!this.llm.tokenize) {
        console.warn("tokenize not implemented for current LLM");
        return [];
      } else {
        return this.llm.tokenize(content);
      }
    });
  }
};
export {
  LLM,
  LLM as LLama
};
