// llama-node/dist/llm/llama-rs.js
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/llm/llama-rs.ts
import {
  EmbeddingResultType,
  InferenceResultType,
  LLama
} from "@llama-node/core";
var LLamaRS = class {
  load(config) {
    this.instance = LLama.create(config);
  }
  createCompletion(params, callback) {
    return __async(this, null, function* () {
      let completed = false;
      const errors = [];
      return new Promise((res, rej) => {
        this.instance.inference(params, (response) => {
          var _a;
          switch (response.type) {
            case InferenceResultType.Data: {
              const data = {
                token: response.data.token,
                completed: !!response.data.completed
              };
              if (data.completed) {
                completed = true;
              }
              callback(data);
              break;
            }
            case InferenceResultType.Error: {
              if (errors.length) {
                rej(new Error(errors.join("\n")));
              } else {
                res(completed);
              }
              break;
            }
            case InferenceResultType.End: {
              errors.push((_a = response.message) != null ? _a : "Unknown error");
              break;
            }
          }
        });
      });
    });
  }
  getEmbedding(params) {
    return __async(this, null, function* () {
      return new Promise((res, rej) => {
        this.instance.getWordEmbeddings(params, (response) => {
          var _a;
          switch (response.type) {
            case EmbeddingResultType.Data:
              res((_a = response.data) != null ? _a : []);
              break;
            case EmbeddingResultType.Error:
              rej(response.message);
              break;
          }
        });
      });
    });
  }
  getDefaultEmbedding(text) {
    return __async(this, null, function* () {
      return this.getEmbedding({
        nThreads: 4,
        numPredict: 1024,
        topK: 40,
        topP: 0.1,
        temp: 0.1,
        repeatPenalty: 1,
        prompt: text
      });
    });
  }
  tokenize(params) {
    return __async(this, null, function* () {
      return new Promise((res) => {
        this.instance.tokenize(params, (response) => {
          res(response.data);
        });
      });
    });
  }
};
export {
  LLamaRS
};
