// llama-node/dist/llm/rwkv-cpp.js
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/llm/rwkv-cpp.ts
import {
  InferenceResultType,
  Rwkv,
  TokenizeResultType
} from "@llama-node/rwkv-cpp";
var RwkvCpp = class {
  load(config) {
    const { modelPath, tokenizerPath, nThreads, enableLogging } = config;
    this.instance = Rwkv.load(
      modelPath,
      tokenizerPath,
      nThreads,
      enableLogging
    );
  }
  createCompletion(params, callback) {
    return __async(this, null, function* () {
      let completed = false;
      const errors = [];
      return new Promise((res, rej) => {
        this.instance.inference(params, (response) => {
          var _a;
          switch (response.type) {
            case InferenceResultType.Data: {
              const data = {
                token: response.data.token,
                completed: !!response.data.completed
              };
              if (data.completed) {
                completed = true;
              }
              callback(data);
              break;
            }
            case InferenceResultType.End: {
              if (errors.length) {
                rej(new Error(errors.join("\n")));
              } else {
                res(completed);
              }
              break;
            }
            case InferenceResultType.Error: {
              errors.push((_a = response.message) != null ? _a : "Unknown Error");
              break;
            }
          }
        });
      });
    });
  }
  // embedding not implemented yet
  /* async getEmbedding(params: RwkvInvocation): Promise<number[]> {
          return new Promise<number[]>((res, rej) => {
              this.instance.getWordEmbedding(params, (response) => {
                  switch (response.type) {
                      case EmbeddingResultType.Data:
                          res(response.data ?? []);
                          break;
                      case EmbeddingResultType.Error:
                          rej(new Error("Unknown Error"));
                          break;
                  }
              });
          });
      }
  
      async getDefaultEmbedding(text: string): Promise<number[]> {
          return this.getEmbedding({
              topP: 0.1,
              temp: 0.1,
              prompt: text,
              maxPredictLength: 2048,
          });
      } */
  tokenize(params) {
    return __async(this, null, function* () {
      return new Promise((res, rej) => {
        this.instance.tokenize(params.content, (response) => {
          if (response.type === TokenizeResultType.Data) {
            res(response.data);
          } else {
            rej(new Error("Unknown Error"));
          }
        });
      });
    });
  }
};
export {
  RwkvCpp
};
